File: GFG POTD/022-September-2025/018-next-greater-element.cpp
// https://www.geeksforgeeks.org/problems/next-greater-element/1/




#include <bits/stdc++.h>
using namespace std;

class Solution {
  public:
    // TC : O(N)
    // SC : O(N)
    vector<int> nextGreater(vector<int> &arr) {
        int n = arr.size();
        vector<int> result(n, -1);
        stack<int> stk;
        for (int i = 2*n - 1; i >= 0; i --) {
            while (!stk.empty() && stk.top() <= arr[i % n]) {
                stk.pop();
            }
            result[i % n] = stk.empty() ? -1 : stk.top();
            stk.push(arr[i % n]);
        }
        return result;
    }
};

File: GFG POTD/022-September-2025/018-next-greater-element.java
// https://www.geeksforgeeks.org/problems/next-greater-element/1/




import java.util.*;

class Solution {
    // TC : O(N)
    // SC : O(N)
    public ArrayList<Integer> nextGreater(int[] arr) {
        int n = arr.length;
        ArrayList<Integer> result = new ArrayList(n);
        for (int i = 0; i < n; i ++) result.add(-1);
        
        Stack<Integer> stk = new Stack<>();
        
        for (int i = 2 * n - 1; i >= 0; i --) {
            while (!stk.isEmpty() && stk.peek() <= arr[i % n]) {
                stk.pop();
            }
            result.set(i % n, (stk.isEmpty() ? -1 : stk.peek()));
            stk.push(arr[i % n]);
        }
        return result;
    }
}